### ISSUE TITLE
additional Build task dependencies listed in $PSBPreference.Build.Dependencies do not run.

### ISSUE THREAD

**Issue by joeypiccola** [psake/PowerShellBuild#37]
I have the task `BumpVersion` that I want the `Build` task to depend on. That said, I found `$PSBPreference.Build.Dependencies` and have supplied the following array `@('StageFiles', 'BuildHelp', 'BumpVersion')`. However, `BumpVersion` never runs. `StageFiles` and `BuildHelp` do run.

## Expected Behavior
When running the `Build` task, additional dependencies listed in `$PSBPreference.Build.Dependencies` should also run.

## Current Behavior
When running the `Build` task, additional dependencies listed in `$PSBPreference.Build.Dependencies` do not run. Strangely enough, when supplying `$PSBPreference.Build.Dependencies` with only `BumpVersion` the default build task dependencies `StageFiles` and `BuildHelp` still run making me think `$PSBPreference.Build.Dependencies` is not even being correctly passed. 

## Possible Solution
I've verified `$PSBPreference.Build.Dependencies` is being supplied to the `Build` task here: https://github.com/psake/PowerShellBuild/blob/11e1b94f22b295cfa7a6494bc7d0c25e2fb52864/PowerShellBuild/psakeFile.ps1#L54

## Steps to Reproduce (for bugs)
1. simply set `$PSBPreference.Build.Dependencies` in the `properties` block of a `psakeFile.ps1` with the value of @('MyExtraBuildTask')
2. create a task in the `psakeFile.ps1` e.g...
```
task MyExtraBuildTask {
    "doing extra things"
}
```
3. run build.ps1

## Context
I have a few extra tasks that I want to run as part of the build. Specifically, a task that bumps the module version if it has not already been incremented and also a task that uploads test results to appveyor when the BuildSystem is appveyor. 

## Your Environment

* Module version used:
PowerShellBuild = 0.3.0
Psake = 4.8.0
* Operating System and PowerShell version:
MacOS 10.14.6
PowerShell = 6.2.0

**Comment by devblackops** [comment]
@joeypiccola tl:dr: It's an ordering issue in the way psake loads the tasks.

When you tell psake to load tasks from a module, it runs the `psakeFile.ps1` provided in that module. That file will load the default properties including the default values in `$PSBPreference.Build.Dependencies`. The tasks in that file are then loaded into the psake context (with whatever attributes are specified). psake **then** loads your `psakeFile.ps1` that overrides `$PSBPreference.Build.Dependencies` but it doesn't matter as those tasks have already been loaded into psake and their attributes are already set.

When you tell psake to load a task from a module, you can **override** things like `-Depends` explicitly.

Try this:

```powershell
task Build -FromModule PowerShellBuild -depends @('StageFiles', 'BuildHelp', 'BumpVersion')

task BumpVersion {
    'Bump version'
}
```

This will give the behavior you want, but ideally we also want to way to do the same when modifying ``$PSBPreference`.

**Comment by joeypiccola** [comment]
@devblackops Thanks for the help, much appreciated. 

> but it doesn't matter as those tasks have already been loaded into psake and their attributes are already set

So what is the intended usage of PowerShellBuilds's `$PSBPreference.Build.Dependencies` if not to append or override the default `Build` task dependencies?

As for what you suggested trying, the following results in the task `testTask` not running. Hopefully, I have understood you correctly in that by running `task Build` I am interacting with the PowerShellBuild's task `Build`. 

```powershell
task default -depends Test
task Test -FromModule PowerShellBuild -Version '0.3.0'
task Build -FromModule PowerShellBuild -Version '0.3.0' -depends @('StageFiles', 'BuildHelp', 'testTask')

task testTask {
    'my test task'
}
```
OUTPUT
```plaintext
Task: BUILD TIME REPORT
Name             Duration
----             --------
Init             00:00:00.170
Clean            00:00:00.008
StageFiles       00:00:00.181
GenerateMarkdown 00:00:02.547
GenerateMAML     00:00:01.219
BuildHelp        00:00:00.002
Build            00:00:00.000
Pester           00:00:07.390
Analyze          00:00:00.214
Test             00:00:00.001
Total:           00:00:11.834
```

**Comment by taalmahret** [comment]
This appears to still remain an issue.  Was there a better way of handling the Shared Task dependencies?

**Comment by jhendricks123** [comment]
I'm having the same issue. My project includes a .NET Framework DLL written in C# which needs to be compiled, and the output staged with the PowerShell module files prior to building the module itself. I was thinking a good way to do this would be to add my task as a dependency to Build but I'm unable to get my task to run when specifying it in $PSBPreference or in the -depends task parameter.

For now, I think I'll work around it by using a separate psakeFile and calling both from build.ps1.

**Comment by rdavisunr** [comment]
@joeypiccola 

Looks like you can change the task import order to make your 2nd example work:

```powershell
task default -depends Test
task Build -FromModule PowerShellBuild -Version '0.6.1' -depends @('StageFiles', 'BuildHelp', 'testTask')
task Test -FromModule PowerShellBuild -Version '0.6.1'

task testTask{
    'my test task'
}
```

Compared to your original (2nd) example, I swapped the order of `task Build` and `task Test` so the `Build` with -depends comes first.

I assume this is related to the same ordering issue originally mentioned. So, a Task with overridden dependencies needs to be specified prior to importing another task that itself depends on the task with overridden dependencies. 

@devblackops - being able to hook custom tasks into the build chain seems like a pretty common use case. Is this the correct method?

**Comment by HeyItsGilbert** [comment]
This should be made easier with #72 which will allow you to change/append additional tasks.

**Comment by AWells1037** [comment]
> This should be made easier with [#72](https://github.com/psake/PowerShellBuild/pull/72) which will allow you to change/append additional tasks.

It looks like even after #72 was merged this isn't completely fixed. I downloaded 0.7.2 because I have the same issue, and I still am not able to override the Publish step. Adapting the sample code from @rdavisunr, the Publish step still attempts to run the Analyze and Pester steps.

```powershell
task default -depends Test
task Build -FromModule PowerShellBuild -Version '0.7.2' -depends @('Init')
task Publish -FromModule PowerShellBuild -Version '0.7.2' -Depends @('Init')

task testTask{
    'my test task'
}
```

The code above in a psake file still runs the tasks 'Pester', and 'Analyze' before the task 'Publish'. The only way I was able to override the publish step was to edit the build.properties.ps1 file in the downloaded module, specifically lines 147-159:

```powershell
TaskDependencies = @{
      Clean = @('Init')
      StageFiles = @('Clean')
      Build = @('StageFiles', 'BuildHelp')
      Analyze = @('Build')
      Pester = @('Build')
      Test = @('Pester', 'Analyze')
      BuildHelp = @('GenerateMarkdown', 'GenerateMAML')
      GenerateMarkdown = @('StageFiles')
      GenerateMAML = @('GenerateMarkdown')
      GenerateUpdatableHelp = @('BuildHelp')
      Publish = @('Test')
}
```

**Comment by HeyItsGilbert** [comment]
I have confirmed the issue still persists. I realized this is due to the fact that the properties (specifically the `$PSBPreference`) are not evaluated before the tasks are loaded. I tested tossing a variable in the `psakeFile.ps1` and updated the module to use that, and that would work. This would likely mean using a different set of variable(s).

I'll noodle on this a bit. There's a chance this could come from a parameter, but I personally wouldn't want to remember to modify that parameter every time. I think a new variable set may be a better option. One that is clearly for "compile" time, vs "run" time.

**Comment by HeyItsGilbert** [comment]
v0.7.3 was published yesterday. @AWells1037 can you give that a go?  See the https://github.com/psake/PowerShellBuild?tab=readme-ov-file#modifying-task-dependencies section on how to implement.  The tl;dr is that you'll need to overwrite the dependencies with a specific variable set within the `psakeFile.ps1` (not the `Properties` block). 

Feel free to re-open if the issue isn't resolved.

**Comment by AWells1037** [comment]
@HeyItsGilbert I downloaded 0.7.3 today and tested the new dependency variables and it looks like everything is working as expected. Thanks for your hard work on this!
